### PR TITLE
Create one bitcode library for AMD

### PR DESCRIPTION
**Pull request by MartinWehking**
Enable compilation of libdevice for AMD by adding AMDGCN to macro
guarded code parts in libdevice for enabling e.g. standard library
math function.

Add compilation workflow to SYCLLibdevice.cmake for AMD.

Follow the compilation mechanism for NVPTX
(https://github.com/intel/llvm/commit/56a6ae2e31f3cf8bd784f3cad5275d9f6981598d)
and create a single bitcode library file.

Do not select builtin LLVM intrinsics for AMDGCN by default to ensure
that stdlib functions can be found when linking against libdevice.

Ensure that the clang tests check for the correctness of the new
clang driver actions and check if the driver still links the device
code against the itt device libraries when device library linkage has
been excluded.

Fix a compilation error of Intel math function libraries for MSVC
when targeting AMD. Include "device.h" before including "device_imf.hpp"
to avoid the inclusion of <type_traits>, which failed with a
redefinition of symbols error.

### CODE REVIEW

**Review by steffenlarsen — APPROVED** [review]
Looks reasonable to me! ⭐

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 1]
```diff
@@ -278,6 +278,9 @@ SYCL::getDeviceLibraries(const Compilation &C, const llvm::Triple &TargetTriple,
   if (TargetTriple.isNVPTX() && IgnoreSingleLibs)
     LibraryList.push_back(Args.MakeArgString("devicelib--cuda.bc"));
```
Comment (frasercrmck) [inline clang/lib/Driver/ToolChains/SYCL.cpp:279]: We should remember to rename these libraries to something more akin to `devicelib-<triple>.bc` or (`sycl-devicelib-<triple>.bc`?) as we discussed

File: libdevice/cmake/modules/SYCLLibdevice.cmake [hunk 2]
```diff
@@ -444,7 +451,7 @@ foreach(dtype IN ITEMS bf16 fp32 fp64)
   endforeach()
 endforeach()
 
-# Add device fallback imf libraries for the CUDA target.
+# Add device fallback imf libraries for the CUDA and AMD targets.
```
Comment (frasercrmck) [inline libdevice/cmake/modules/SYCLLibdevice.cmake:454]: We seem to be mixing terminology here, which is understandable and it happens often, but `CUDA` and `AMD` stand out to me here. Either NVPTX/AMD or CUDA/HIP?

**Review by frasercrmck — APPROVED** [review]

File: clang/lib/Driver/ToolChains/SYCL.cpp (hunk 1 above)
Comment (MartinWehking) [reply clang/lib/Driver/ToolChains/SYCL.cpp:279]: That's a good point. We might as well do this in this patch already for AMD.
I can put up another one for the NVPTX one as well

**Review by elizabethandrews — APPROVED** [review]
FE changes LGTM

**Review by smanna12 — APPROVED** [review]

File: clang/lib/Driver/ToolChains/SYCL.cpp (hunk 1 above)
Comment (MartinWehking) [reply clang/lib/Driver/ToolChains/SYCL.cpp:279]: After a bit of consideration, I would leave it as is for now and put up a separate patch with those renamings for the NVPTX and AMD libaries later.

CMake renamings could cause dependency errors and in that case this entire patch would have to be reverted (As you pointed out)

**Comment by MartinWehking** [comment]
ping @intel/dpcpp-clang-driver-reviewers

File: clang/lib/Driver/ToolChains/SYCL.cpp (hunk 1 above)
Comment (frasercrmck) [reply clang/lib/Driver/ToolChains/SYCL.cpp:279]: I agree :+1:

File: clang/test/Driver/sycl-device-lib-amdgcn.cpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
+// Tests specific to `-fsycl-targets=amdgcn-amd-amdhsa`
+// Verify that the correct devicelib linking actions are spawned by the driver.
+// Check also if the correct warnings are generated.
+
+// UNSUPPORTED: system-windows
+// REQUIRES: amdgpu-registered-target
```
Comment (mdtoguchi) [inline clang/test/Driver/sycl-device-lib-amdgcn.cpp:6]: The use of a `REQUIRES` target shouldn't be necessary with driver specific tests.  The use of the `--sysroot=%S/Inputs/SYCL` will allow for local dummy files to be picked up to verify driver functionality.  Adding an empty `Inputs/SYCL/lib/devicelib--amd.bc` file should help this along.

Comment (MartinWehking) [reply clang/test/Driver/sycl-device-lib-amdgcn.cpp]: Ah, I see. Thanks for clarifying this.
I've removed this and added a dummy input

**Review by mdtoguchi — APPROVED** [review]
LGTM - thanks!

**Comment by AlexeySachkov** [comment]
@ldrumm, it seems like this PR introduced some new failures on AMD in post-commit, could you please take a look?

https://github.com/intel/llvm/actions/runs/11031105056/job/30637542157

It also looks like we have some problems with build log extraction, because logs below contain some random symbols as if we read logs from incorrect memory location.

```
********************
FAIL: SYCL :: Basic/launch_queries/max_num_work_groups.cpp (294 of 2185)
******************** TEST 'SYCL :: Basic/launch_queries/max_num_work_groups.cpp' FAILED ********************
Exit Code: -6

Command Output (stdout):
--
# RUN: at line 1
/__w/llvm/llvm/toolchain/bin//clang++  -Werror -Xsycl-target-backend=amdgcn-amd-amdhsa --offload-arch=gfx1031 -fsycl -fsycl-targets=amdgcn-amd-amdhsa  /__w/llvm/llvm/llvm/sycl/test-e2e/Basic/launch_queries/max_num_work_groups.cpp -o /__w/llvm/llvm/build-e2e/Basic/launch_queries/Output/max_num_work_groups.cpp.tmp.out
# executed command: /__w/llvm/llvm/toolchain/bin//clang++ -Werror -Xsycl-target-backend=amdgcn-amd-amdhsa --offload-arch=gfx1031 -fsycl -fsycl-targets=amdgcn-amd-amdhsa /__w/llvm/llvm/llvm/sycl/test-e2e/Basic/launch_queries/max_num_work_groups.cpp -o /__w/llvm/llvm/build-e2e/Basic/launch_queries/Output/max_num_work_groups.cpp.tmp.out
# note: command had no output on stdout or stderr
# RUN: at line 2
env ONEAPI_DEVICE_SELECTOR=hip:gpu  /__w/llvm/llvm/build-e2e/Basic/launch_queries/Output/max_num_work_groups.cpp.tmp.out
# executed command: env ONEAPI_DEVICE_SELECTOR=hip:gpu /__w/llvm/llvm/build-e2e/Basic/launch_queries/Output/max_num_work_groups.cpp.tmp.out
# .---command stderr------------
# | <HIP>[ERROR]: 
# | UR HIP ERROR:
# | 	Value:           209
# | 	Name:            hipErrorNoBinaryForGpu
# | 	Description:     no kernel image is available for execution on the device
# | 	Function:        buildProgram
# | 	Source Location: /__w/llvm/llvm/build/_deps/unified-runtime-src/source/adapters/hip/program.cpp:235
# | 
# | terminate called after throwing an instance of 'sycl::_V1::exception'
# |   what():  The program was built for 1 devices
# | Build program log for 'AMD Radeon RX 6700 XT':
# | PT
# `-----------------------------
# error: command failed with exit status: -6

--

********************
FAIL: SYCL :: Basic/kernel_bundle/kernel_bundle_api_hip.cpp (373 of 2185)
******************** TEST 'SYCL :: Basic/kernel_bundle/kernel_bundle_api_hip.cpp' FAILED ********************
Exit Code: -6

Command Output (stdout):
--
# RUN: at line 2
/__w/llvm/llvm/toolchain/bin//clang++  -Werror -Xsycl-target-backend=amdgcn-amd-amdhsa --offload-arch=gfx1031 -fsycl -fsycl-targets=amdgcn-amd-amdhsa  /__w/llvm/llvm/llvm/sycl/test-e2e/Basic/kernel_bundle/kernel_bundle_api_hip.cpp -fsycl-device-code-split=per_kernel -o /__w/llvm/llvm/build-e2e/Basic/kernel_bundle/Output/kernel_bundle_api_hip.cpp.tmp.out
# executed command: /__w/llvm/llvm/toolchain/bin//clang++ -Werror -Xsycl-target-backend=amdgcn-amd-amdhsa --offload-arch=gfx1031 -fsycl -fsycl-targets=amdgcn-amd-amdhsa /__w/llvm/llvm/llvm/sycl/test-e2e/Basic/kernel_bundle/kernel_bundle_api_hip.cpp -fsycl-device-code-split=per_kernel -o /__w/llvm/llvm/build-e2e/Basic/kernel_bundle/Output/kernel_bundle_api_hip.cpp.tmp.out
# note: command had no output on stdout or stderr
# RUN: at line 3
env ONEAPI_DEVICE_SELECTOR=hip:gpu  /__w/llvm/llvm/build-e2e/Basic/kernel_bundle/Output/kernel_bundle_api_hip.cpp.tmp.out
# executed command: env ONEAPI_DEVICE_SELECTOR=hip:gpu /__w/llvm/llvm/build-e2e/Basic/kernel_bundle/Output/kernel_bundle_api_hip.cpp.tmp.out
# .---command stderr------------
# | Empty list of devices for get_kernel_bundle
# | Caught: Not all devices are associated with the context or vector of devices is empty
# | Expect: Not all devices are associated with the context or vector of devices is empty
# | Mismatched contexts for join
# | <HIP>[ERROR]: 
# | UR HIP ERROR:
# | 	Value:           209
# | 	Name:            hipErrorNoBinaryForGpu
# | 	Description:     no kernel image is available for execution on the device
# | 	Function:        buildProgram
# | 	Source Location: /__w/llvm/llvm/build/_deps/unified-runtime-src/source/adapters/hip/program.cpp:235
# | 
# | <HIP>[ERROR]: 
# | UR HIP ERROR:
# | 	Value:           UR_RESULT_ERROR_UNKNOWN
# | 	Function:        urProgramCompile
# | 	Source Location: /__w/llvm/llvm/build/_deps/unified-runtime-src/source/adapters/hip/program.cpp:286
# | 
# | Caught: The program was built for 1 devices
# | Build program log for 'AMD Radeon RX 6700 XT':
# | �է
# | Expect: Not all input bundles have the same associated context
# | kernel_bundle_api_hip.cpp.tmp.out: /__w/llvm/llvm/llvm/sycl/test-e2e/Basic/kernel_bundle/kernel_bundle_api_hip.cpp:25: void checkException(TryBodyT, const std::string &) [TryBodyT = (lambda at /__w/llvm/llvm/llvm/sycl/test-e2e/Basic/kernel_bundle/kernel_bundle_api_hip.cpp:168:9)]: Assertion `CorrectException && "Test failed: caught exception is incorrect."' failed.
# `-----------------------------
# error: command failed with exit status: -6

--

********************
FAIL: SYCL :: GroupAlgorithm/root_group.cpp (1236 of 2185)
******************** TEST 'SYCL :: GroupAlgorithm/root_group.cpp' FAILED ********************
Exit Code: -6

Command Output (stdout):
--
# RUN: at line 3
/__w/llvm/llvm/toolchain/bin//clang++  -Werror -Xsycl-target-backend=amdgcn-amd-amdhsa --offload-arch=gfx1031 -fsycl -fsycl-targets=amdgcn-amd-amdhsa  /__w/llvm/llvm/llvm/sycl/test-e2e/GroupAlgorithm/root_group.cpp -I . -o /__w/llvm/llvm/build-e2e/GroupAlgorithm/Output/root_group.cpp.tmp.out
# executed command: /__w/llvm/llvm/toolchain/bin//clang++ -Werror -Xsycl-target-backend=amdgcn-amd-amdhsa --offload-arch=gfx1031 -fsycl -fsycl-targets=amdgcn-amd-amdhsa /__w/llvm/llvm/llvm/sycl/test-e2e/GroupAlgorithm/root_group.cpp -I . -o /__w/llvm/llvm/build-e2e/GroupAlgorithm/Output/root_group.cpp.tmp.out
# note: command had no output on stdout or stderr
# RUN: at line 4
env ONEAPI_DEVICE_SELECTOR=hip:gpu  /__w/llvm/llvm/build-e2e/GroupAlgorithm/Output/root_group.cpp.tmp.out
# executed command: env ONEAPI_DEVICE_SELECTOR=hip:gpu /__w/llvm/llvm/build-e2e/GroupAlgorithm/Output/root_group.cpp.tmp.out
# .---command stderr------------
# | <HIP>[ERROR]: 
# | UR HIP ERROR:
# | 	Value:           209
# | 	Name:            hipErrorNoBinaryForGpu
# | 	Description:     no kernel image is available for execution on the device
# | 	Function:        buildProgram
# | 	Source Location: /__w/llvm/llvm/build/_deps/unified-runtime-src/source/adapters/hip/program.cpp:235
# | 
# | terminate called after throwing an instance of 'sycl::_V1::exception'
# |   what():  The program was built for 1 devices
# | Build program log for 'AMD Radeon RX 6700 XT':
# | �#e�(
# `-----------------------------
# error: command failed with exit status: -6

--

********************
FAIL: SYCL :: KernelAndProgram/kernel-bundle-find-run.cpp (1[59](https://github.com/intel/llvm/actions/runs/11031105056/job/30637542157#step:22:60)7 of 2185)
******************** TEST 'SYCL :: KernelAndProgram/kernel-bundle-find-run.cpp' FAILED ********************
Exit Code: -6

Command Output (stdout):
--
# RUN: at line 1
/__w/llvm/llvm/toolchain/bin//clang++  -Werror -Xsycl-target-backend=amdgcn-amd-amdhsa --offload-arch=gfx1031 -fsycl -fsycl-targets=amdgcn-amd-amdhsa  /__w/llvm/llvm/llvm/sycl/test-e2e/KernelAndProgram/kernel-bundle-find-run.cpp -o /__w/llvm/llvm/build-e2e/KernelAndProgram/Output/kernel-bundle-find-run.cpp.tmp.out
# executed command: /__w/llvm/llvm/toolchain/bin//clang++ -Werror -Xsycl-target-backend=amdgcn-amd-amdhsa --offload-arch=gfx1031 -fsycl -fsycl-targets=amdgcn-amd-amdhsa /__w/llvm/llvm/llvm/sycl/test-e2e/KernelAndProgram/kernel-bundle-find-run.cpp -o /__w/llvm/llvm/build-e2e/KernelAndProgram/Output/kernel-bundle-find-run.cpp.tmp.out
# note: command had no output on stdout or stderr
# RUN: at line 2
env ONEAPI_DEVICE_SELECTOR=hip:gpu  /__w/llvm/llvm/build-e2e/KernelAndProgram/Output/kernel-bundle-find-run.cpp.tmp.out
# executed command: env ONEAPI_DEVICE_SELECTOR=hip:gpu /__w/llvm/llvm/build-e2e/KernelAndProgram/Output/kernel-bundle-find-run.cpp.tmp.out
# .---command stdout------------
# | sycl_kernel done
# `-----------------------------
# .---command stderr------------
# | <HIP>[ERROR]: 
# | UR HIP ERROR:
# | 	Value:           209
# | 	Name:            hipErrorNoBinaryForGpu
# | 	Description:     no kernel image is available for execution on the device
# | 	Function:        buildProgram
# | 	Source Location: /__w/llvm/llvm/build/_deps/unified-runtime-src/source/adapters/hip/program.cpp:235
# | 
# | terminate called after throwing an instance of 'sycl::_V1::exception'
# |   what():  The program was built for 1 devices
# | Build program log for 'AMD Radeon RX [67](https://github.com/intel/llvm/actions/runs/11031105056/job/30637542157#step:22:68)00 XT':
# | �G�
# `-----------------------------
# error: command failed with exit status: -6

--

********************
FAIL: SYCL :: KernelAndProgram/kernel-bundle-get-kernel.cpp (1599 of 2185)
******************** TEST 'SYCL :: KernelAndProgram/kernel-bundle-get-kernel.cpp' FAILED ********************
Exit Code: -6

Command Output (stdout):
--
# RUN: at line 1
/__w/llvm/llvm/toolchain/bin//clang++  -Werror -Xsycl-target-backend=amdgcn-amd-amdhsa --offload-arch=gfx1031 -fsycl -fsycl-targets=amdgcn-amd-amdhsa  /__w/llvm/llvm/llvm/sycl/test-e2e/KernelAndProgram/kernel-bundle-get-kernel.cpp -o /__w/llvm/llvm/build-e2e/KernelAndProgram/Output/kernel-bundle-get-kernel.cpp.tmp.out
# executed command: /__w/llvm/llvm/toolchain/bin//clang++ -Werror -Xsycl-target-backend=amdgcn-amd-amdhsa --offload-arch=gfx1031 -fsycl -fsycl-targets=amdgcn-amd-amdhsa /__w/llvm/llvm/llvm/sycl/test-e2e/KernelAndProgram/kernel-bundle-get-kernel.cpp -o /__w/llvm/llvm/build-e2e/KernelAndProgram/Output/kernel-bundle-get-kernel.cpp.tmp.out
# note: command had no output on stdout or stderr
# RUN: at line 2
env ONEAPI_DEVICE_SELECTOR=hip:gpu  /__w/llvm/llvm/build-e2e/KernelAndProgram/Output/kernel-bundle-get-kernel.cpp.tmp.out
# executed command: env ONEAPI_DEVICE_SELECTOR=hip:gpu /__w/llvm/llvm/build-e2e/KernelAndProgram/Output/kernel-bundle-get-kernel.cpp.tmp.out
# .---command stderr------------
# | <HIP>[ERROR]: 
# | UR HIP ERROR:
# | 	Value:           209
# | 	Name:            hipErrorNoBinaryForGpu
# | 	Description:     no kernel image is available for execution on the device
# | 	Function:        buildProgram
# | 	Source Location: /__w/llvm/llvm/build/_deps/unified-runtime-src/source/adapters/hip/program.cpp:235
# | 
# | terminate called after throwing an instance of 'sycl::_V1::exception'
# |   what():  The program was built for 1 devices
# | Build program log for 'AMD Radeon RX 6[70](https://github.com/intel/llvm/actions/runs/11031105056/job/30637542157#step:22:71)0 XT':
# | 
# `-----------------------------
# error: command failed with exit status: -6

--

********************
FAIL: SYCL :: SpecConstants/2020/host_apis.cpp (1903 of 2185)
******************** TEST 'SYCL :: SpecConstants/2020/host_apis.cpp' FAILED ********************
Exit Code: -6

Command Output (stdout):
--
# RUN: at line 1
/__w/llvm/llvm/toolchain/bin//clang++  -Werror -Xsycl-target-backend=amdgcn-amd-amdhsa --offload-arch=gfx1031 -fsycl -fsycl-targets=amdgcn-amd-amdhsa  /__w/llvm/llvm/llvm/sycl/test-e2e/SpecConstants/2020/host_apis.cpp -Wno-error=unused-command-line-argument -o /__w/llvm/llvm/build-e2e/SpecConstants/2020/Output/host_apis.cpp.tmp.out -fsycl-dead-args-optimization
# executed command: /__w/llvm/llvm/toolchain/bin//clang++ -Werror -Xsycl-target-backend=amdgcn-amd-amdhsa --offload-arch=gfx1031 -fsycl -fsycl-targets=amdgcn-amd-amdhsa /__w/llvm/llvm/llvm/sycl/test-e2e/SpecConstants/2020/host_apis.cpp -Wno-error=unused-command-line-argument -o /__w/llvm/llvm/build-e2e/SpecConstants/2020/Output/host_apis.cpp.tmp.out -fsycl-dead-args-optimization
# .---command stderr------------
# | clang++: warning: argument unused during compilation: '-fsycl-dead-args-optimization' [-Wunused-command-line-argument]
# `-----------------------------
# RUN: at line 2
env ONEAPI_DEVICE_SELECTOR=hip:gpu  /__w/llvm/llvm/build-e2e/SpecConstants/2020/Output/host_apis.cpp.tmp.out
# executed command: env ONEAPI_DEVICE_SELECTOR=hip:gpu /__w/llvm/llvm/build-e2e/SpecConstants/2020/Output/host_apis.cpp.tmp.out
# .---command stderr------------
# | <HIP>[ERROR]: 
# | UR HIP ERROR:
# | 	Value:           209
# | 	Name:            hipErrorNoBinaryForGpu
# | 	Description:     no kernel image is available for execution on the device
# | 	Function:        buildProgram
# | 	Source Location: /__w/llvm/llvm/build/_deps/unified-runtime-src/source/adapters/hip/program.cpp:235
# | 
# | terminate called after throwing an instance of 'sycl::_V1::exception'
# |   what():  The program was built for 1 devices
# | Build program log for 'AMD Radeon RX 6700 XT':
# | @
# `-----------------------------
# error: command failed with exit status: -6

--
```

**Comment by sarnex** [comment]
@frasercrmck @MartinWehking Can someone take a look at Alexey's comment about this change breaking HIP CI? Thanks.

**Comment by frasercrmck** [comment]
> @frasercrmck @MartinWehking Can someone take a look at Alexey's comment about this change breaking HIP CI? Thanks.

Yep we're taking a look :+1:

**Comment by npmiller** [comment]
This patch should fix all of the failures:
* https://github.com/intel/llvm/pull/15558

We'll need to follow up on this to figure out what's going on with the build log, but it should be enough to unblock the CI

**Pull request merged**